models/esss.R

```diff
@@ -0,0 +1,111 @@
+# ESSS
+#
+# Model "ESSS" is a flexible exponential smoothing state space model
+#  fit using the forecast and ets functions with the possibilit of 
+#  multiplicative trends. Unfortunately because of the seasonality and 
+#  sampling occurring with different frequencies, which the ets function 
+#  cannot accommodate, seasonal models are not included.
+
+  library(forecast)
+
+#' Function for ESSS
+#'
+#' @param abundances table of rodent abundances and time measures
+#' @param forecast_date the dates to be forecast from
+#' @param forecast_months the months of the dates to be forecast
+#' @param forecast_years the years of the dates to be forecast
+#' @param forecast_newmoons the numbers of the new moons to be forecast
+#' @param level name of the type of plots included ("All" or "Controls")
+#' @param num_forecast_newmoons number of new moons to forecast
+#' @param CI_level confidence interval level used for forecast envelope
+#' @return list of forecast and aic tables
+
+  forecast_esss <- function(abundances, forecast_date, forecast_months, 
+                            forecast_years, forecast_newmoons, level,
+                            num_forecast_newmoons = 12, CI_level = 0.9){
+
+    cat("Fitting ESSS model for total \n")
+    interpolated_abundances <- interpolate_abundance(abundances)
+
+    ets_model <- ets(interpolated_abundances$total)
+    ets_forecast <- forecast(ets_model, h = num_forecast_newmoons,
+                             level = CI_level, 
+                             allow.multiplicative.trend = TRUE)
+
+    fit_start_newmoon <- min(abundances$newmoonnumber)
+    fit_end_newmoon <- max(abundances$newmoonnumber)
+    initial_newmoon <- max(abundances$newmoonnumber)
+    CI_match <- which(ets_forecast$level == CI_level * 100)
+
+    output_fcast <- data.frame(date = forecast_date, 
+                               forecastmonth = forecast_months,
+                               forecastyear = forecast_years, 
+                               newmoonnumber = forecast_newmoons,
+                               currency = "abundance",
+                               model = "ESSS", 
+                               level = level, 
+                               species = "total", 
+                               estimate = ets_forecast$mean,
+                               LowerPI = ets_forecast$lower[ , CI_match], 
+                               UpperPI = ets_forecast$upper[ , CI_match],
+                               fit_start_newmoon = fit_start_newmoon,
+                               fit_end_newmoon = fit_end_newmoon,
+                               initial_newmoon = initial_newmoon)
+    for(i in 5:8)
+      output_fcast[ , i] <- as.character(output_fcast[ , i])
+    for(i in 9:11)
+      output_fcast[ , i] <- as.numeric(output_fcast[ , i])
```

Comment:
Does this conversion from time-series to numeric influence the output/warnings? If not then I'd drop this. If so then I'd probably add the `as.numeric`'s into the data frame definition instead. The idea being that

```
estimate = as.numeric(ets_forecast$mean)
```

reads pretty clearly as "Define estimate as ets_forecast$mean while addressing a type issue", where as this loop requires more thought - "OK, they just defined a data frame, and they are looping over it to do what... OK, changing type... So, which columns are 9:11..."

Reply:
yeah, it's to change the predictions from class ts to numeric
I like your suggestion here, although to avoid long lines, I'd suggest we pull the definition out to before the data frame definition

`estimate <- as.numeric(ets_forecast$mean)`
`output_fcast <- data.frame(..., estimate = estimate, ...)`

let me know if you'd see any issues with that

Reply:
👍